### PR TITLE
Add the compiler guide

### PR DESCRIPTION
**Pull request by cartermp**
First stab at incorporating these docs into the repo.

### CODE REVIEW

File: docs/compiler-guide.md [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+
+For example, commands like Find All References and Rename can be cheap if a codebase is small, hence the lower bound being `S`. But if the symbol in question is used across many documents in a large project graph, they are very expensive because the entire graph must be crawled and all symbols contained in its documents must be inspected.
+
+In contrast, actions like highlighting all symbols in a document aren't terribly expensive even for very large file files. That's because the symbols to be inspected are ultimately only in a single document.
+
+Operations that use typecheck data execute on a single background thread (see [Reactor.fsi](https://github.com/dotnet/fsharp/blob/master/src/fsharp/service/Reactor.fsi)/[Reactor.fs](https://github.com/dotnet/fsharp/blob/master/src/fsharp/service/Reactor.fs)). Each operation is cancellable - for example, if you run an expensive Find All References but decide to do something else, the next action you take that requires semantic data will cancel the Find All References operation and start the one you requested.
+
+TODO for don --> why single-threaded? why no priority queue if can't be multithreaded?
```
Comment (cartermp) [inline docs/compiler-guide.md:465]: @dsyme I need your help explaining why the reactor is single-threaded

File: docs/compiler-guide.md [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+
+## `eventually` computations
+
+Some parts of the F# codebase (specifically, the type checker) are written using `eventually` computation expressions. These define resumption-like computations which can be  time-sliced, suspended or discarded at "bind" points.
+
+This is done to ensure that long-running type-checking and other computations in the F# Compiler Service can be interrupted and cancelled. The documentation of the [F# Compiler Service Operations Queue](fsharp-compiler-service-queue.md) covers some aspects of this.
+
+When compiling code with `fsc` or executing code with `fsi`, these computations are not time-sliced and simply run synchronously and without interruption (unless the user requests it).
+
+TODO for don --> is the below stuff still accurate?
```
Comment (cartermp) [inline docs/compiler-guide.md:534]: @dsyme I need your help checking if the words in the rest of this section are still accurate

File: docs/compiler-guide.md [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+
+The first can be interrupted by having the incremental builder dependency graph
+(see [IncrementalBuild.fsi](https://github.com/fsharp/FSharp.Compiler.Service/tree/master/src/fsharp/service/IncrementalBuild.fsi)/[IncrementalBuild.fs](https://github.com/fsharp/FSharp.Compiler.Service/tree/master/src/fsharp/service/IncrementalBuild.fs))
+decide not to bother continuing with the computation (it drops it on the floor)
+
+The second can be interrupted via having `isResultObsolete` to the F# Compiler Service API return true.
+
+### The F# Compiler Service Operations Queue
+
+See [F# Compiler Service Queue](fsharp-compiler-service-queue.md).
```
Comment (cartermp) [inline docs/compiler-guide.md:549]: This needs to change

File: docs/compiler-guide.md [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+
+### The F# Compiler Service Caches
+
+See [F# Compiler Service Caches](fsharp-compiler-service-caches.md).
```
Comment (cartermp) [inline docs/compiler-guide.md:553]: This needs to change

File: docs/compiler-guide.md [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+
+Compiler startup performance is a key factor affecting happiness of F# users. If the compiler took 10sec to start up, then far fewer people would use F#.
+
+On all platforms, the following factors affect startup performance:
+
+* Time to load compiler binaries. This depends on the size of the generated binaries, whether they are pre-compiled (for example, using NGEN or CrossGen), and the way the .NET implementation loads them.
+
+* Time to open referenced assemblies (for example, `mscorlib.dll`, `FSharp.Core.dll`) and analyze them for the types and namespaces defined. This depends particularly on whether this is correctly done in an on-demand way.
+
+* Time to process "open" declarations are the top of each file. Processing these declarations have been observed to take time in some cases of  F# compilation.
```
Comment (deviousasti) [inline docs/compiler-guide.md:372]: *at* the top

**Comment by deviousasti** [comment]
Should documentation about expression elaboration be here?

File: docs/compiler-guide.md (hunk 2 above)
Comment (dsyme) [reply docs/compiler-guide.md:534]: I think we should integrate now and then later today I do an end to end edit as a separate PR

Comment (cartermp) [reply docs/compiler-guide.md:534]: Okay. Can you approve?

**Review by KevinRansom — APPROVED** [review]

**Comment by cartermp** [comment]
Will merge this to get it in now, there are some things that could likely get improved and we'll address those independently I suppose

**Pull request merged**